jstz_api/src/urlpattern/mod.rs

```diff
@@ -0,0 +1,576 @@
+            "username",
+            get:((url, context) => Ok(url.username().into_js(context)))
+        )
+    }
+
+    fn test(
+        this: &JsValue,
+        args: &[JsValue],
+        context: &mut Context<'_>,
+    ) -> JsResult<JsValue> {
+        let url_pattern = UrlPattern::try_from_js(this)?;
+        let input: UrlPatternInput = match args.get(0) {
+            Some(value) => value.try_js_into(context)?,
+            None => UrlPatternInput::default(),
+        };
+        let base_url: Option<String> = args.get_or_undefined(1).try_js_into(context).ok();
+        Ok(url_pattern.test(input, base_url)?.into_js(context))
+    }
+
+    fn exec(
+        this: &JsValue,
+        args: &[JsValue],
+        context: &mut Context<'_>,
+    ) -> JsResult<JsValue> {
+        let url_pattern = UrlPattern::try_from_js(this)?;
+        let input: UrlPatternInput = match args.get(0) {
+            Some(value) => value.try_js_into(context)?,
+            None => UrlPatternInput::default(),
+        };
+        let base_url: Option<String> = args.get_or_undefined(1).try_js_into(context).ok();
+        url_pattern
+            .exec(input, base_url)?
+            .map_or(Ok(JsValue::Null), |r| Ok(r.into_js(context)))
+    }
+}
+
+impl TryFromJs for UrlPatternInit {
+    fn try_from_js(value: &JsValue, context: &mut Context<'_>) -> JsResult<Self> {
+        let obj = value.as_object().ok_or_else(|| {
+            JsError::from_native(JsNativeError::typ().with_message("Expected `JsObject`"))
+        })?;
+
+        macro_rules! get_optional_property {
+            ($obj:ident, $field:literal, $context:ident) => {
+                if $obj.has_property(js_string!($field), $context)? {
+                    $obj.get(js_string!($field), $context)?
+                        .try_js_into($context)?
+                } else {
+                    None
+                }
+            };
+        }
+
+        let url_pattern_init = urlpattern::quirks::UrlPatternInit {
+            protocol: get_optional_property!(obj, "protocol", context),
+            username: get_optional_property!(obj, "username", context),
+            password: get_optional_property!(obj, "password", context),
+            hostname: get_optional_property!(obj, "hostname", context),
+            port: get_optional_property!(obj, "port", context),
+            pathname: get_optional_property!(obj, "pathname", context),
+            search: get_optional_property!(obj, "search", context),
+            hash: get_optional_property!(obj, "hash", context),
+            base_url: get_optional_property!(obj, "base_url", context),
+        };
+
+        Ok(Self(url_pattern_init))
+    }
+}
+
+impl TryFromJs for UrlPatternInput {
+    fn try_from_js(value: &JsValue, context: &mut Context<'_>) -> JsResult<Self> {
+        if let Some(string) = value.as_string() {
+            return Ok(Self(InnerStringOrQuirksInit::String(
+                string.to_std_string_escaped(),
+            )));
+        };
```

Comment:
Try to always use existing converters over destructuring
```suggestion
        if value.is_string() {
            let string: String = value.try_js_into(context)?;
            return Ok(Self(InnerStringOrQuirksInit::String(string)))
        };
```

Reply:
Applied!